orttraining/orttraining/training_api/optimizer.cc

```diff
@@ -5,58 +5,94 @@
+        ORT_ENFORCE(OrtValueLike(optim_sess_state, pair.second->Data(), param_state).IsOK());
+        cur_param_optimizer_states.momentum_named_states.insert({state_name, std::move(param_state)});
       }
     }
   }
+
+  std::shared_ptr<onnxruntime::Model> model;
+  ORT_THROW_IF_ERROR(onnxruntime::Model::Load(optim_path_or_bytes, model, nullptr, env->GetLoggingManager()->DefaultLogger()));
+  GetGraphInputOutputNames(model->MainGraph(), input_names_, output_names_);
+  ORT_ENFORCE(input_names_[0] == "learning_rate");  // TODO: make this better
+  ORT_ENFORCE(input_names_[1] == "step");           // TODO: make this better
+
+  std::string param_name;
+  std::vector<std::string> param_names, grad_names, moment1_names, moment2_names, user_inputs;
+  for (size_t i = 2; i < input_names_.size(); i++) {
+    std::string& name = input_names_[i];
+    auto it = parameters_.find(name);
+    if (it != parameters_.end()) {  // is param
+      param_names.push_back(name);
+      inputs_.push_back(it->second->Data());
+    } else if (GetParamNameFromGradient(name, param_name)) {
+      grad_names.emplace_back(name);
+      // assert param_name is valid.
+      auto it = parameters_.find(param_name);
+      ORT_ENFORCE(it != parameters_.end(), "Unknown param: ", param_name, " for field: ", name);
+      inputs_.push_back(it->second->Gradient());
+    } else if (GetParamNameFromSuffix(name, MOMENT_1_SUFFIX, param_name)) {
+      moment1_names.push_back(name);
+      auto it = parameters_.find(param_name);
+      ORT_ENFORCE(it != parameters_.end(), "Unknown param: ", param_name, " for field: ", name);
+      inputs_.push_back(param_named_optimizer_states.at(param_name).momentum_named_states.at(MOMENT_STATE_NAMES[0]));
+    } else if (GetParamNameFromSuffix(name, MOMENT_2_SUFFIX, param_name)) {
+      moment2_names.push_back(name);
+      auto it = parameters_.find(param_name);
+      ORT_ENFORCE(it != parameters_.end(), "Unknown param: ", param_name, " for field: ", name);
+      inputs_.push_back(param_named_optimizer_states.at(param_name).momentum_named_states.at(MOMENT_STATE_NAMES[1]));
+    } else {
+      ORT_THROW("This is an invalid graph. Optimizer graph contains unknown user input:",name);
+    }
+  }
+}
+
+Status Optimizer::Step() {
+  OrtValue learning_rate_input, step_input;
+  WarpInOrtValue<float>(optimizer_state_.learning_rate, &learning_rate_input);
+  WarpInOrtValue<int64_t>(optimizer_state_.step, &step_input);
+  std::vector<OrtValue> feeds({learning_rate_input, step_input});
+  feeds.insert(feeds.end(), inputs_.begin(), inputs_.end());
+
+  std::vector<OrtValue> outputs;
+  auto status = optim_sess_->Run(RunOptions(), input_names_, feeds, output_names_, &outputs);
+  ORT_THROW_IF_ERROR(status);
+
+  // extract step output and update
+  // TODO: need to remove hardcoding
+  optimizer_state_.step = GetValue<int64_t>(outputs[0]);
```

Comment:
why not just step++, getValue() here means, we have a synchronize here to get the CPU copy of the value from CUDA

Reply:
the step output is generated on CPU, so theres no memcpy